// discord-ginko/src/Structures/Comando.js
const GinkoError = require('../Util/GinkoError')

class Comando {

    constructor(options = {}){ 

        this.run = options.run

        this.nombre = options.nombre

        this.descripcion = options.descripcion

        this.categoria = options.categoria

        this.uso = options.uso 

        this.soloOwner = options.soloOwner || false

        this.cooldown = options.cooldown || false

        this.alias = options.alias || false

        if(!options.permisoRequerido) this.permiso = false
        else{ 
        
        if(typeof options.permisoRequerido !== 'object') throw new GinkoError('permisoRequerido tiene que ser un objeto!')
        for (var i in options.permisoRequerido){
            if(i !== 'permisos' && i !== 'respuesta') throw new GinkoError('[ '+ i +' ] No es una Key valida para el objeto!')
            if(i == 'permisos' && !(options.permisoRequerido[i] instanceof Array)) throw new GinkoError('Los permisos tienen que estar en un Array')
        }
        this.permiso = options.permisoRequerido
        }
    }

}
module.exports = Comando;
